lib/cdo/crowdin/utils.rb

```diff
@@ -0,0 +1,86 @@
+    #  written out in JSON format
+    # @param options.locales_dir [String, nil] path to directory where changed
+    #  files should be downloaded
+    # @param options.logger [Logger, nil]
+    def initialize(project, options={})
+      @project = project
+      @changes_json = options.fetch(:changes_json, "/tmp/#{project.id}_changes.json")
+      @etags_json = options.fetch(:etags_json, "/tmp/#{project.id}_etags.json")
+      @locales_dir = options.fetch(:locales_dir, "/tmp/locales")
+      @logger = options.fetch(:logger, Logger.new(STDOUT))
+    end
+
+    # Fetch from Crowdin a list of files changed since the last sync. Uses
+    # etags sourced from the @etags_json file to define what we mean by "since
+    # the last sync," and writes the results out to @changes_json.
+    def fetch_changes
+      etags = File.exist?(@etags_json) ? JSON.parse(File.read(@etags_json)) : {}
+      changes = {}
+
+      languages = @project.languages
+      num_languages = languages.length
+      languages.each_with_index do |language, i|
+        code = language["code"]
+        @logger.debug("#{language['name']} (#{code}): #{i}/#{num_languages}")
+        @logger.info("~#{(i * 100 / num_languages).round(-1)}% complete (#{i}/#{num_languages})") if i > 0 && i % (num_languages / 5) == 0
+
+        etags[code] ||= {}
+        files = @project.list_files
+
+        results = Parallel.map(files) do |file|
+          etag = etags[code].fetch(file, nil)
+          response = @project.export_file(file, code, etag)
+          case response.code
+          when 200
+            [file, response.headers["etag"]]
+          when 304
+            nil
+          else
+            raise "cannot handle response code #{response.code}"
+          end
+        end.compact
+
+        next if results.empty?
+
+        changes[code] = results.to_h
+        etags[code].merge!(changes[code])
+        File.write(@etags_json, JSON.pretty_generate(etags))
+        File.write(@changes_json, JSON.pretty_generate(changes))
+      end
+
+      File.write(@changes_json, JSON.pretty_generate(changes))
+    end
+
+    # Downloades all files referenced in @changes_json to @locales_dir
+    def download_changed_files
+      changes = JSON.parse(File.read(@changes_json))
+      @logger.info("#{changes.keys.length} languages have changes")
+      @project.languages.each do |language|
+        code = language["code"]
+        name = language["name"]
+        files = changes.fetch(code, nil)
+        next unless files.present?
+        filenames = files.keys
+        locale_dir = File.join(@locales_dir, name)
+
+        @logger.debug("#{name} (#{code}): #{filenames.length} files have changes")
+        Parallel.each(filenames) do |file|
+          response = @project.export_file(file, code)
+          dest = File.join(locale_dir, file)
+          FileUtils.mkdir_p(File.dirname(dest))
+          File.write(dest, response.body)
+        end
+      end
+    end
```

Comment:
> Note that we could just download the changed files directly, rather than first generating a list of changes and then downloading the files in that list. That would make this operation somewhat simpler and slightly faster, but I plan to also use that list of changed files to dramatically speed up the sync out so I think it's worth it to build this in a more complex way right off the bat.

It looks like [the export-file api](https://support.crowdin.com/api/export-file/) _does_ download the changed files, even if the If-None-Match header is provided.  That would mean we are downloading the file [when we call `export_file` with the `etag` param in `fetch_changes`](https://github.com/code-dot-org/code-dot-org/pull/34846/files#diff-92acf60ab8570fb67a0108e82c0b49dfR42). So in practice, I think we are downloading the changed files twice, we're just throwing away the response the first time.  Is there an optimization I've missed that avoids this?

I don't see yet why you couldn't generate a list of changed files and download them at the same time, but still keep that list around for future optimizations.

Reply:
oh, good call! I'll update the `export_file` call to optionally use `head` rather than `get`

Reply:
Clever! I'm guessing that works great, even though their documentation doesn't discuss `head` requests.  Good idea.